File: backend/app/schemas/video.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field

from app.schemas.production import ProductionBriefResponse


class VideoBase(BaseModel):
    """Base video schema with common fields."""

    filename: str
    duration: Optional[float] = None
    file_size: Optional[int] = None
    resolution: Optional[str] = None
    fps: Optional[float] = None
    codec: Optional[str] = None
    created_date: Optional[datetime] = None


class VideoCreate(VideoBase):
    """Schema for creating a new video."""

    file_path: str


class VideoUpdate(BaseModel):
    """Schema for updating video metadata."""

    category: Optional[str] = None
    location: Optional[str] = None
    notes: Optional[str] = None
    tags: Optional[list[str]] = None
    production_ids: Optional[list[int]] = None


class MetadataResponse(BaseModel):
    """Schema for metadata in video responses."""

    category: Optional[str] = None
    location: Optional[str] = None
    notes: Optional[str] = None

    class Config:
        from_attributes = True


class TagResponse(BaseModel):
    """Schema for tag in video responses."""

    id: int
    name: str

    class Config:
        from_attributes = True


class VideoResponse(VideoBase):
    """Schema for video response with all related data."""

    id: int
    file_path: str
    indexed_date: datetime
    thumbnail_count: int
    metadata: Optional[MetadataResponse] = Field(None, alias='video_metadata')
    tags: list[TagResponse] = []
    productions: list[ProductionBriefResponse] = []

    class Config:
        from_attributes = True
        populate_by_name = True


class VideoListResponse(BaseModel):
    """Schema for paginated video list response."""

    videos: list[VideoResponse]
    total: int
    page: int
    limit: int
    pages: int


class BulkUpdateRequest(BaseModel):
    """Schema for bulk update request."""

    video_ids: list[int] = Field(..., min_length=1)
    category: Optional[str] = None
    location: Optional[str] = None
    notes: Optional[str] = None
    add_tags: Optional[list[str]] = None
    remove_tags: Optional[list[str]] = None
    add_production_ids: Optional[list[int]] = None
    remove_production_ids: Optional[list[int]] = None


class BulkUpdateResponse(BaseModel):
    """Schema for bulk update response."""

    updated: int
    message: str

File: backend-rust/src/models/video.rs
use chrono::NaiveDateTime;
use diesel::prelude::*;
use serde::{Deserialize, Serialize};

use crate::schema::{videos, metadata};

#[derive(Debug, Queryable, Selectable, Identifiable, Serialize, Clone)]
#[diesel(table_name = videos)]
pub struct Video {
    pub id: i32,
    pub file_path: String,
    pub filename: String,
    pub duration: Option<f32>,
    pub file_size: Option<i64>,
    pub resolution: Option<String>,
    pub fps: Option<f32>,
    pub codec: Option<String>,
    pub created_date: Option<NaiveDateTime>,
    pub indexed_date: NaiveDateTime,
    pub thumbnail_count: i32,
    pub checksum: Option<String>,
}

#[derive(Debug, Insertable)]
#[diesel(table_name = videos)]
pub struct NewVideo {
    pub file_path: String,
    pub filename: String,
    pub duration: Option<f32>,
    pub file_size: Option<i64>,
    pub resolution: Option<String>,
    pub fps: Option<f32>,
    pub codec: Option<String>,
    pub created_date: Option<NaiveDateTime>,
    pub indexed_date: NaiveDateTime,
    pub thumbnail_count: i32,
    pub checksum: Option<String>,
}

#[derive(Debug, Queryable, Selectable, Identifiable, Associations, Serialize, Clone)]
#[diesel(table_name = metadata)]
#[diesel(belongs_to(Video, foreign_key = video_id))]
pub struct Metadata {
    pub id: i32,
    pub video_id: i32,
    pub category: Option<String>,
    pub location: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Insertable)]
#[diesel(table_name = metadata)]
pub struct NewMetadata {
    pub video_id: i32,
    pub category: Option<String>,
    pub location: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, AsChangeset)]
#[diesel(table_name = metadata)]
pub struct MetadataChangeset {
    pub category: Option<String>,
    pub location: Option<String>,
    pub notes: Option<String>,
}

/// Full video data with all relationships loaded
#[derive(Debug, Serialize, Clone)]
pub struct VideoFull {
    pub id: i32,
    pub file_path: String,
    pub filename: String,
    pub duration: Option<f32>,
    pub file_size: Option<i64>,
    pub resolution: Option<String>,
    pub fps: Option<f32>,
    pub codec: Option<String>,
    pub created_date: Option<NaiveDateTime>,
    pub indexed_date: NaiveDateTime,
    pub thumbnail_count: i32,
    pub checksum: Option<String>,
    pub metadata: Option<MetadataResponse>,
    pub tags: Vec<TagResponse>,
    pub productions: Vec<ProductionBriefResponse>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct MetadataResponse {
    pub category: Option<String>,
    pub location: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TagResponse {
    pub id: i32,
    pub name: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ProductionBriefResponse {
    pub id: i32,
    pub title: String,
    pub platform: Option<String>,
    pub link: Option<String>,
    pub is_published: bool,
}

#[derive(Debug, Deserialize)]
pub struct VideoUpdate {
    pub category: Option<String>,
    pub location: Option<String>,
    pub notes: Option<String>,
    pub tags: Option<Vec<String>>,
    pub production_ids: Option<Vec<i32>>,
}

#[derive(Debug, Serialize)]
pub struct VideoListResponse {
    pub videos: Vec<VideoFull>,
    pub total: i64,
    pub page: i64,
    pub limit: i64,
    pub pages: i64,
}

#[derive(Debug, Deserialize)]
pub struct BulkUpdateRequest {
    pub video_ids: Vec<i32>,
    pub category: Option<String>,
    pub location: Option<String>,
    pub notes: Option<String>,
    pub add_tags: Option<Vec<String>>,
    pub remove_tags: Option<Vec<String>>,
    pub add_production_ids: Option<Vec<i32>>,
    pub remove_production_ids: Option<Vec<i32>>,
}

#[derive(Debug, Serialize)]
pub struct BulkUpdateResponse {
    pub updated: i64,
    pub message: String,
}
